app/react/src/client/preview/utils.ts

```diff
@@ -0,0 +1,58 @@
+// TODO: add client-api as dependency
+import { defaultDecorateStory, combineParameters } from '@storybook/client-api';
+import { GlobalConfig, Meta, Story, StoryContext } from './types-6-0';
+
+function composeStory<GenericArgs>(
+  meta: Meta,
+  globalConfig: GlobalConfig,
+  story: Story<GenericArgs>
+) {
+  const finalStoryFn = (context: StoryContext) => {
+    const { passArgsFirst = true } = context.parameters;
+    return passArgsFirst ? story(context.args as GenericArgs, context) : story(context);
+  };
+
+  const combinedDecorators = [
+    ...(globalConfig?.decorators || []),
+    ...(meta?.decorators || []),
+    ...(story.decorators || story.story?.decorators || []),
+  ];
```

Comment:
FYI, I copy/pasted this functionality (`composeStory`) to run it locally, and the decorators are applied in the wrong order. They should be structured: Global > Meta > Story, but passing this array to `defaultDecorateStory` results in the opposite: Story > Meta > Global.

```suggestion
  const combinedDecorators = [
    ...(story.decorators || story.story?.decorators || []),
    ...(meta?.decorators || []),
    ...(globalConfig?.decorators || []),
  ];
```

---

As an aside, I'm using `composeStory` to render fully-decorated stories outside of Storybook (in a Next.js app). Exporting this function, in addition to the likely more commonly used `composeStories` would be helpful for that purpose. Thanks!

Reply:
Hey @kylegach thanks for the tips, I've updated the code accordingly and also exposed `composeStory`! Could you show me what you mean by using this to render fully decorated stories in a Next.js app? Is there another scenario you're using this rather than in unit tests?

Reply:
You're welcome, Yann. Thanks for exposing that function.

For my project, I'm basically trying to recreate the MDX docs experience within Next.js. We use Storybook to define the stories themselves (in pure CSF), which provides infrastructure for tests and a component dev environment. But rather than write MDX in Storybook, I'm trying to use Next.js, which mostly gives us total control over the design and structure of the site (e.g. splitting navigation into primary and secondary, instead of one big sidebar tree forced by Storybook). MDX is also very important from an authorship standpoint, allowing non-technical contributors. To do that, I've modified Next's webpack config to consume the stories and TypeScript docgen info (the Storybook and Next.js projects are different packages in a monorepo), and I need those stories to be fully-decorated when I render them on the page. I've got that working quite well, with the exception of static assets, which I suspect will require more webpack configuration.

Admittedly, that's a lot of complexity for not much reason, but it fits our needs well. It's all in a private repo, but if you really need me to, I may be able to recreate a simple version in a public sample repo.

That's most of it, but happy to discuss on Discord if you'd like to know more.

Reply:
Woah that's quite a setup hehe. Glad this has helped you out.
 I'll reach you out on discord in case I have further questions! thanks for checking this PR and I hope you're having great holidays!